presto-hive/src/test/java/com/facebook/presto/hive/util/TestAsyncQueue.java

```diff
@@ -0,0 +1,258 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.hive.util;
+
+import com.facebook.presto.hive.util.AsyncQueue.BorrowResult;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.concurrent.Threads;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.util.ArrayList;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static io.airlift.testing.Assertions.assertContains;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.fail;
+
+public class TestAsyncQueue
+{
+    private ExecutorService executor;
+
+    @BeforeClass
+    public void setUpClass()
+            throws Exception
+    {
+        executor = Executors.newFixedThreadPool(8, Threads.daemonThreadsNamed("test-async-queue-%s"));
+    }
+
+    @AfterClass
+    public void tearDownClass()
+            throws Exception
+    {
+        executor.shutdownNow();
+
+    }
+
+    @Test(timeOut = 10_000)
+    public void testGetPartial()
+            throws Exception
+    {
+        AsyncQueue<String> queue = new AsyncQueue<>(4, executor);
+
+        queue.offer("1");
+        queue.offer("2");
+        queue.offer("3");
+        assertEquals(queue.getBatchAsync(100).get(), ImmutableList.of("1", "2", "3"));
+
+        queue.finish();
+        assertTrue(queue.isFinished());
```

Comment:
verify get after finish and maybe?

Reply:
This is tested in `testOfferAfterFinish`